File: benchmark/mnist.py
from __future__ import print_function
from torchvision import datasets, transforms
import os
import pathlib
import time
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(28 * 28, 512)
        self.fc2 = nn.Linear(512, 512)
        self.fc3 = nn.Linear(512, 10)

    def forward(self, x):
        x = x.view((-1, 28 * 28,))
        x = self.fc1(x)
        x = torch.tanh(x)
        x = self.fc2(x)
        x = torch.tanh(x)
        x = self.fc3(x)
        return F.log_softmax(x, dim=1)


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('Test set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

def main():
    device = torch.device("cpu")

    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
    ])

    cache_dir = os.path.join(str(pathlib.Path.home()), ".cache/mnist")
    dataset1 = datasets.MNIST(cache_dir, train=True, download=True,
                             transform=transform)
    dataset2 = datasets.MNIST(cache_dir, train=False,
                       transform=transform)

    train_loader = torch.utils.data.DataLoader(dataset1, batch_size=64)
    test_loader =torch.utils.data.DataLoader(dataset2, batch_size=64)

    model = Net().to(device)
    model.train()
    optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.5)

    epochs = 5
    for epoch in range(epochs):
        start = time.time()
        for batch_idx, (data, target) in enumerate(train_loader):
            data, target = data.to(device), target.to(device)
            optimizer.zero_grad()
            output = model(data)
            loss = F.nll_loss(output, target)
            loss.backward()
            optimizer.step()
        throughput = len(dataset1) * 1.0 / (time.time() - start)
        print('Train Epoch: {}, Loss: {:.6f}, throughput: {} samples/sec'.format(
                epoch, loss.item(), throughput))
        test(model, device, test_loader)


if __name__ == '__main__':
    main()

File: example/mnist/mnist.go
package main

import (
	"encoding/gob"
	"flag"
	"fmt"
	"image"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	torch "github.com/wangkuiyi/gotorch"
	F "github.com/wangkuiyi/gotorch/nn/functional"
	"github.com/wangkuiyi/gotorch/nn/initializer"
	"github.com/wangkuiyi/gotorch/vision/datasets"
	"github.com/wangkuiyi/gotorch/vision/models"
	"github.com/wangkuiyi/gotorch/vision/transforms"
)

var device torch.Device

func main() {
	if torch.IsCUDAAvailable() {
		log.Println("CUDA is valid")
		device = torch.NewDevice("cuda")
	} else {
		log.Println("No CUDA found; CPU only")
		device = torch.NewDevice("cpu")
	}

	initializer.ManualSeed(1)

	trainCmd := flag.NewFlagSet("train", flag.ExitOnError)
	trainTar := trainCmd.String("data", "/tmp/mnist_png_training_shuffled.tar.gz", "data tarball")
	testTar := trainCmd.String("test", "/tmp/mnist_png_testing_shuffled.tar.gz", "data tarball")
	save := trainCmd.String("save", "/tmp/mnist_model.gob", "the model file")
	epoch := trainCmd.Int("epoch", 5, "the number of epochs")

	predictCmd := flag.NewFlagSet("predict", flag.ExitOnError)
	load := predictCmd.String("load", "/tmp/mnist_model.gob", "the model file")

	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s needs subcomamnd train or predict\n", os.Args[0])
		os.Exit(1)
	}

	switch os.Args[1] {
	case "train":
		trainCmd.Parse(os.Args[2:])
		train(*trainTar, *testTar, *epoch, *save)
	case "predict":
		predictCmd.Parse(os.Args[2:])
		predict(*load, predictCmd.Args())
	}
}

func train(trainFn, testFn string, epochs int, save string) {
	vocab, e := datasets.BuildLabelVocabularyFromTgz(trainFn)
	if e != nil {
		panic(e)
	}
	net := models.MLP()
	net.To(device)
	opt := torch.SGD(0.01, 0.5, 0, 0, false)
	opt.AddParameters(net.Parameters())
	defer torch.FinishGC()

	for epoch := 0; epoch < epochs; epoch++ {
		var trainLoss float32
		startTime := time.Now()
		trainLoader := MNISTLoader(trainFn, vocab)
		testLoader := MNISTLoader(testFn, vocab)
		totalSamples := 0
		for trainLoader.Scan() {
			data, label := trainLoader.Minibatch()
			totalSamples += int(data.Shape()[0])
			opt.ZeroGrad()
			pred := net.Forward(data.To(device, data.Dtype()))
			loss := F.NllLoss(pred, label.To(device, label.Dtype()), torch.Tensor{}, -100, "mean")
			loss.Backward()
			opt.Step()
			trainLoss = loss.Item().(float32)
		}
		throughput := float64(totalSamples) / time.Since(startTime).Seconds()
		log.Printf("Train Epoch: %d, Loss: %.4f, throughput: %f samples/sec", epoch, trainLoss, throughput)
		test(net, testLoader)
	}
	saveModel(net, save)
}

// MNISTLoader returns a ImageLoader with MNIST training or testing tgz file
func MNISTLoader(fn string, vocab map[string]int) *datasets.ImageLoader {
	trans := transforms.Compose(transforms.ToTensor(), transforms.Normalize([]float64{0.1307}, []float64{0.3081}))
	loader, e := datasets.NewImageLoader(fn, vocab, trans, 64)
	if e != nil {
		panic(e)
	}
	return loader
}

func test(model *models.MLPModule, loader *datasets.ImageLoader) {
	testLoss := float32(0)
	correct := int64(0)
	samples := 0
	for loader.Scan() {
		data, label := loader.Minibatch()
		data = data.To(device, data.Dtype())
		label = label.To(device, label.Dtype())
		output := model.Forward(data)
		loss := F.NllLoss(output, label, torch.Tensor{}, -100, "mean")
		pred := output.Argmax(1)
		testLoss += loss.Item().(float32)
		correct += pred.Eq(label.View(pred.Shape()...)).Sum(map[string]interface{}{"dim": 0, "keepDim": false}).Item().(int64)
		samples += int(label.Shape()[0])
	}
	log.Printf("Test average loss: %.4f, Accuracy: %.2f%%\n",
		testLoss/float32(samples), 100.0*float32(correct)/float32(samples))
}

func saveModel(model *models.MLPModule, modelFn string) {
	log.Println("Saving model to", modelFn)
	f, e := os.Create(modelFn)
	if e != nil {
		log.Fatalf("Cannot create file to save model: %v", e)
	}
	defer f.Close()

	d := torch.NewDevice("cpu")
	model.To(d)
	if e := gob.NewEncoder(f).Encode(model.StateDict()); e != nil {
		log.Fatal(e)
	}
}

func predict(modelFn string, inputs []string) {
	net := loadModel(modelFn)

	for _, in := range inputs {
		for _, pa := range strings.Split(in, ":") {
			fns, e := filepath.Glob(pa)
			if e != nil {
				log.Fatal(e)
			}

			for _, fn := range fns {
				predictFile(fn, net)
			}
		}
	}
}

func loadModel(modelFn string) *models.MLPModule {
	f, e := os.Open(modelFn)
	if e != nil {
		log.Fatal(e)
	}
	defer f.Close()

	states := make(map[string]torch.Tensor)
	if e := gob.NewDecoder(f).Decode(&states); e != nil {
		log.Fatal(e)
	}

	net := models.MLP()
	net.SetStateDict(states)
	return net
}

func predictFile(fn string, m *models.MLPModule) {
	f, e := os.Open(fn)
	if e != nil {
		log.Fatal(e)
	}
	defer f.Close()

	img, _, e := image.Decode(f)
	if e != nil {
		log.Fatalf("Cannot decode input image: %v", e)
	}

	t := transforms.ToTensor().Run(img)
	n := transforms.Normalize([]float64{0.1307}, []float64{0.3081}).Run(t)
	fmt.Println(m.Forward(n).Argmax().Item())
}
